arrays/arrays.c

```diff
@@ -77,30 +81,43 @@ void resize_array(Array *arr) {
 char *arr_read(Array *arr, int index) {
 
   // Throw an error if the index is greater than the current count
-  if(index > arr->count) {
-      printf("Out of Index Range");
+  if(index >= arr->count) {
+      printf("Out of Index Range\n");
       return NULL;
   }
   // Otherwise, return the element at the given index
   //printf("Arr[%d]: %s\n", index, arr->elements[index]);
   return arr->elements[index];
 }
 
-
 /*****
  * Insert an element to the array at the given index
  *****/
 void arr_insert(Array *arr, char *element, int index) {
 
   // Throw an error if the index is greater than the current count
+  if(index < 0 || index > arr->capacity - 1) {
+      printf("Index out of Range\n");
+      return;
+  }
 
   // Resize the array if the number of elements is over capacity
+  if(arr->count + 1 > arr->capacity) {
+      printf("Array Doubling from %d to %d...\n", arr->capacity, arr->capacity*2);
+      resize_array(arr);
+  }
 
   // Move every element after the insert index to the right one position
+  for(int i = arr->capacity-1; i >= index; i--) {
+      arr->elements[i+1] = arr->elements[i];
+  }
+
 
   // Copy the element and add it to the array
+  arr->elements[index] = element;
```

Comment:
Similar to arr_append, you should make a copy of element and then assign elements[index] to element.